llvm/lib/Target/RISCV/RISCVInstrInfo.cpp

```diff
@@ -629,6 +629,21 @@ bool RISCVInstrInfo::isAsCheapAsAMove(const MachineInstr &MI) const {
   switch(Opcode) {
     default:
       break;
+    case RISCV::CMove:
+      return true;
+    case RISCV::CIncOffset:
+      // Incrementing a capability by zero should be as fast as a move since
+      // it is effectively the same thing (other special-casing sealed inputs).
```

Comment:
CHERI-RISC-V doesn't have the `cincoffset cd, cs, zero` special case that MIPS has and will trap.

Reply:
MIPS also traps now since the special case was removed. There is a missing "than" though.
And yes, they are not the same thing as moves and could be more expensive.
Unfortunately LLVM does not rematerialize unless "isAsCheapAsAMove" also returns true, so we miss out on real code-generation improvements.

Returning true here should not result in worse code generation anywhere since isAsCheapAsAMove is not really used for anything other than sinking/hoisting/rematerialization decisions. If you can come up with a counter-example, I'll happily drop this change.

Reply:
FYI, if you look at AArch64InstrInfo::isAsCheapAsAMove(), they return it for almost all arithmetic operations so maybe it should be "isAlmostAsCheapAsAMove"?.

Reply:
Do we ever expect to generate a CIncOffset with zero/0? I'd hope that, if it's pointer arithmetic, the constant 0 would have been removed by the optimiser turning it into a COPY/CMove, and since we don't even have the intrinsic any more and the builtin is just a GEP that should be true even then?

Reply:
As you can see in https://github.com/CTSRD-CHERI/llvm-project/pull/471/commits/fcad0052d1491aa967ebc58c4cf5dc307f5943a9, the code-genration for the byval-buffer.ll/byval-varargs.ll tests only improves after adding this case. I added some debug prints and it turns out we generate a CIncOffsetImm 0. I'm not sure where it's being created but my guess would be prolog/epilog.

Reply:
Hmm, `llvm-objdump -d ~/cheri/build/qtwebkit-riscv64-purecap-build/bin/DumpRenderTree | awk '/cincoffset/ && ($NF == "0" || $NF == "zero"){print}'` turns up a _lot_ of `cincoffset cd, csp, 0` instructions, so yeah we're not using a CMove for the first stack slot.